### ISSUE TITLE
Ambiguity warning: false positive?

### ISSUE THREAD

**Issue by amartgon** [JuliaLang/julia#10174]
In the example bellow I get this warning, even though it seems to me that the two methods are not ambiguous. Am I missing something or is the compiler?

```
myf(x::SparseMatrixCSC, y::SparseMatrixCSC...) = 0
myf(x::Any, y::Array...) = 1

#OUTPUT:

#Warning: New definition
#    myf(Any,Array{T,N}...) at /tmp/test.jl:6
#is ambiguous with:
#    myf(Base.SparseMatrix.SparseMatrixCSC{Tv,Ti<:Integer},Base.SparseMatrix.SparseMatrixCSC{Tv,Ti<:Integer}...) at /tmp/test.jl:5.
#To fix, define
#    myf(Base.SparseMatrix.SparseMatrixCSC{Tv,Ti<:Integer},)
#before the new definition.
```

**Comment by JeffBezanson** [comment]
You're right, the warning does not look necessary.

**Comment by yuyichao** [comment]
I think this ambiguity is real? None of the two definition seems to be more specific than the other one.

**Comment by toivoh** [comment]
I agree with Yichao.

**Comment by timholy** [comment]
Still a problem on current master.

**Comment by yuyichao** [comment]
What's the problem?

**Comment by timholy** [comment]
Only one case:

```jl
julia> S = sprand(3,3,0.4)
3×3 sparse matrix with 3 Float64 nonzero entries:
        [1, 1]  =  0.470274
        [3, 2]  =  0.0942641
        [3, 3]  =  0.214061

julia> @test myf(S) == 0
Error During Test
  Test threw an exception of type MethodError
  Expression: myf(S) == 0
  MethodError: myf(::SparseMatrixCSC{Float64,Int64}) is ambiguous. Candidates:
    svec(Tuple{#myf,SparseMatrixCSC{Float64,Int64}},svec(),myf(x, y::Array...) at REPL[16]:1)
    svec(Tuple{#myf,SparseMatrixCSC{Float64,Int64}},svec(),myf(x::SparseMatrixCSC, y::SparseMatrixCSC...) at REPL[15]:2)
   in eval(::Module, ::Any) at ./boot.jl:230
ERROR: There was an error during testing
 in record(::Base.Test.FallbackTestSet, ::Base.Test.Error) at ./test.jl:321
 in do_test(::Base.Test.Threw, ::Expr) at ./test.jl:219
 in eval(::Module, ::Any) at ./boot.jl:230

julia> @test myf(S, S) == 0
Test Passed
  Expression: myf(S,S) == 0
   Evaluated: 0 == 0

julia> @test myf("foo") == 1
Test Passed
  Expression: myf("foo") == 1
   Evaluated: 1 == 1

julia> @test myf("foo", zeros(2,2)) == 1
Test Passed
  Expression: myf("foo",zeros(2,2)) == 1
   Evaluated: 1 == 1

julia> @test myf(S, zeros(2,2)) == 1
Test Passed
  Expression: myf(S,zeros(2,2)) == 1
   Evaluated: 1 == 1
```

**Comment by yuyichao** [comment]
I guess I still think this is indeed ambiguous... not sure which way is better....

**Comment by JeffBezanson** [comment]
Fixed in 0.6. These aren't ambiguous --- the overlap is `myf(xx::SparseMatrixCSC)`, for which the first definition is obviously more specific.
